File: releasefiles/reports2/policy_violations.py
import r2
import json
import csv


jsonfile = 'policy_violations.json'
csvfile = 'policy_violations.csv'


def makePolicyViolationsReport():
    policies = r2.getNexusIqData('policies')
    policyIds = getPolicyIds(policies)

    policyViolations = r2.getNexusIqData("policyViolations?" + policyIds)

    with open(jsonfile, 'w') as fd:
            json.dump(policyViolations, fd)

    writeToCsvFile(policyViolations)

    print(jsonfile)
    print(csvfile)

    return


def getPolicyIds(data):
    policyIds = ""
    policies = data['policies']

    for policy in policies:
        name = policy["name"]
        id = policy["id"]

        if name in r2.securityPolicyList or name in r2.licensePolicyList:
            policyIds += "p=" + id + "&"

    result = policyIds.rstrip('&')

    return result


def writeToCsvFile(policyViolations):
    applicationViolations = policyViolations['applicationViolations']

    with open(csvfile, 'w') as fd:
            writer = csv.writer(fd)

            line = []
            line.append("PolicyName")
            line.append("Reason")
            line.append("ApplicationName")
            line.append("OpenTime")
            line.append("Component")
            line.append("Stage")

            writer.writerow(line)

            for applicationViolation in applicationViolations:
                applicationPublicId = applicationViolation["application"]["publicId"]

                policyViolations = applicationViolation["policyViolations"]
                for policyViolation in policyViolations:
                    stage = policyViolation["stageId"]
                    openTime = policyViolation["openTime"]
                    policyName = policyViolation["policyName"]
                    packageUrl = policyViolation["component"]["packageUrl"]

                    constraintViolations = policyViolation["constraintViolations"]

                    for constraintViolation in constraintViolations:
                        reason = ""

                        reasons = constraintViolation["reasons"]

                        if policyName == "Integrity-Rating":
                            reason = "Integrity-Rating"
                        elif policyName in r2.securityPolicyList:
                            reason = r2.getCVE(reasons)
                        elif policyName in r2.licensePolicyList:
                            reason = r2.getLicense(reasons)
                        else:
                            reason = ""

                        line = []
                        line.append(policyName)
                        line.append(reason)
                        line.append(applicationPublicId)
                        line.append(openTime)
                        line.append(packageUrl)
                        line.append(stage)

                        writer.writerow(line)

    fd.close()

    return

File: get-metrics/src/main/java/org/sonatype/cs/getmetrics/reports/PolicyViolations.java
package org.sonatype.cs.getmetrics.reports;

import org.apache.commons.lang3.NotImplementedException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.sonatype.cs.getmetrics.service.CsvFileService;
import org.sonatype.cs.getmetrics.service.FileIoService;
import org.sonatype.cs.getmetrics.service.PolicyIdsService;
import org.sonatype.cs.getmetrics.util.FilenameInfo;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import javax.json.JsonArray;
import javax.json.JsonObject;
import javax.json.JsonReader;

public class PolicyViolations implements CsvFileService {
    private static final Logger log = LoggerFactory.getLogger(PolicyViolations.class);

    private final PolicyIdsService policyIdsService;

    public PolicyViolations(PolicyIdsService policyIdsService) {
        this.policyIdsService = policyIdsService;
    }

    @Override
    public void makeCsvFile(FileIoService fileIoService, JsonReader reader) {
        log.info("Make Policy Violations Report");

        List<String[]> data = getPolicyViolationsFromData(reader);

        fileIoService.writeCsvFile(FilenameInfo.POLICY_VIOLATIONS_CSV_FILE, data);
    }

    static List<String[]> getPolicyViolationsFromData(JsonReader reader) {
        List<String[]> data = new ArrayList<>();
        data.add(new String[]{"policyName", "reason", "applicationName", "openTime", "component", "stage", "threatLevel"});

        JsonObject obj = reader.readObject();
        JsonArray results = obj.getJsonArray("applicationViolations");

        for (JsonObject result : results.getValuesAs(JsonObject.class)) {
            JsonObject application = result.getJsonObject("application");
            JsonArray policyViolations = result.getJsonArray("policyViolations");

            String applicationPublicId = application.getString("publicId", "");

            for (JsonObject policyViolation : policyViolations.getValuesAs(JsonObject.class)) {
                String stage = policyViolation.getString("stageId", "");
                String openTime = policyViolation.getString("openTime", "");
                String policyName = policyViolation.getString("policyName", "");
                int threatLevel = policyViolation.getInt("threatLevel");

                JsonObject component = policyViolation.getJsonObject("component");
                String packageUrl = component.getString("packageUrl", "");

                JsonArray constraintViolations = policyViolation.getJsonArray("constraintViolations");
                for (JsonObject constraintViolation : constraintViolations.getValuesAs(JsonObject.class)) {
                    JsonArray reasons = constraintViolation.getJsonArray("reasons");

                    String reason = getReason(policyName, reasons);

                    String[] line = {policyName, reason, applicationPublicId, openTime, packageUrl, stage, String.valueOf(threatLevel)};
                    data.add(line);
                }
            }
        }
        return data;
    }

    private static String getReason(String policyName, JsonArray reasons) {
        if (policyName.equalsIgnoreCase("Integrity-Rating")) {
            return "Integrity-Rating";
        } else if (PolicyIdsService.isSecurityPolicy(policyName)) {
            return getCVE(reasons);
        } else if (PolicyIdsService.isLicensePolicy(policyName)) {
            return getLicense(reasons);
        }
        return "";
    }

    static String getCVE(JsonArray reasons) {
        List<String> cves = new ArrayList<>();
        for (JsonObject reason : reasons.getValuesAs(JsonObject.class)) {
            if (!reason.isNull("reference")) {
                JsonObject reference = reason.getJsonObject("reference");
                String cve = reference.getString("value", "");
                if (!cves.contains(cve)) {
                    cves.add(cve);
                }
            }
        }
        return String.join(":", cves);
    }

    static String getLicense(JsonArray reasons) {
        List<String> licenses = new ArrayList<>();
        Pattern pattern = Pattern.compile("^.+\'(.+)\'");

        for (JsonObject reason : reasons.getValuesAs(JsonObject.class)) {
            String licenseFound = reason.getString("reason", "");
            Matcher matcher = pattern.matcher(licenseFound);
            if (matcher.find()) {
                licenseFound = matcher.group(1);
            }
            if (!licenses.contains(licenseFound)) {
                licenses.add(licenseFound);
            }
        }
        return String.join(":", licenses);
    }

    @Override
    public void makeCsvFile(FileIoService fileIoService, JsonObject reader) {
        throw new NotImplementedException();
    }
}
